### PR TITLE
feat: replace `fast-safe-stringify` with `safe-stable-stringify`

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
 - add `safe-stable-stringify` package

## Context
- usedto get sort json objects before stringifying 
- blocks #16661 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
i though we also replace the existing package

**Comment by RahulGautamSingh** [comment]
> 

It is used aat one place not sure what the use is:
https://github.com/renovatebot/renovate/blob/ab94b9503aaa3c28d523d9c7de6d06cf20b2d84f/lib/util/clone.ts#L4

If we can achieve the same with this package then the exhange is possible.

**Comment by viceice** [comment]
> > 
> 
> The other package is used at one place only and I am not sure what it is used for my guess is sorting config as well:
> 
> https://github.com/renovatebot/renovate/blob/ab94b9503aaa3c28d523d9c7de6d06cf20b2d84f/lib/util/clone.ts#L4
> 
> If we can achieve the same with this package then the exhange is possible;

should work, but we probably need to disable key sorting there.

**Review by rarkins — APPROVED** [review]

**Review by rarkins — COMMENTED** [review]
We used the fast-safe-stringify package because JSON.stringify() hit circular references sometimes, somehow. 

If this replacement reorders objects and is used anywhere we process config then it can reorder config in a problematic way

**Comment by RahulGautamSingh** [comment]
https://github.com/BridgeAR/safe-stable-stringify#stringifyconfigureoptions
We can configure the `stringify` method to not reorder (already done). 
About circular reference I am not sure which value we want to set them to, these are the possible ones:
```
circularValue {string|null|undefined|ErrorConstructor} Defines the value for circular references. Set to undefined, circular properties are not serialized (array entries are replaced with null). Set to Error, to throw on circular references. Default: [Circular].
```

**Comment by rarkins** [comment]
null for circular is ok

**Review by rarkins — DISMISSED** [review]

**Comment by viceice** [comment]
> We can configure the `stringify` method to not reorder (already done). About circular reference I am not sure which value we want to set them to, these are the possible ones:
> 
> ```
> circularValue {string|null|undefined|ErrorConstructor} 
> Defines the value for circular references. 
> Set to undefined, circular properties are not serialized (array entries are replaced with null). 
> Set to Error, to throw on circular references. 
> Default: [Circular].
> ```
> 
> https://github.com/BridgeAR/safe-stable-stringify#stringifyconfigureoptions

the old `fast-safe-stringify` uses same default value, so we should leave it for now

**Review by viceice — DISMISSED** [review]

File: lib/util/clone.spec.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  it('maintains same order', () => {
+    const res = clone(obj);
+    expect(Object.keys(res)).toEqual(Object.keys(obj));
+  });
+
+  it('assigns "[Circular]" to circular references', () => {
+    obj.circular = obj;
+    const res = clone(obj);
+    expect(res.circular).toBe('[Circular]');
```
Comment (rarkins) [inline lib/util/clone.spec.ts:18]: I expected this to be `null`. Is it the same with fast-safe-stringify?

Comment (RahulGautamSingh) [reply lib/util/clone.spec.ts:18]: Yes

**Review by rarkins — APPROVED** [review]

File: lib/util/clone.spec.ts (hunk 1 above)
Comment (viceice) [reply lib/util/clone.spec.ts:18]: https://www.npmjs.com/package/fast-safe-stringify#Usage

**Review by viceice — APPROVED** [review]

**Pull request merged**